test/distributed/_sharded_tensor/test_megatron_prototype.py

```diff
@@ -0,0 +1,119 @@
+        sharded_linear_first_layer.bias = torch.nn.Parameter(local_linear_first_layer.bias.detach().clone())
+        sharded_linear_second_layer.weight = torch.nn.Parameter(local_linear_second_layer.weight.detach().clone())
+        sharded_linear_second_layer.bias = torch.nn.Parameter(local_linear_second_layer.bias.detach().clone())
+
+        # Shard the parameter. First col-wise sharding and then row-wise
+        shard_parameter(sharded_linear_first_layer, "weight", spec[0])
+        shard_parameter(sharded_linear_second_layer, "weight", spec[1])
+        act_func = torch.nn.GELU()
+
+        # Run sharded computation
+        torch.manual_seed(self.rank)  # inputs different on each rank
+        inp = torch.rand(*input_size).cuda(self.rank)
+        temp = sharded_linear_first_layer(inp)
+        temp2 = act_func(sharded_linear_first_layer(inp))
+        sharded_output = sharded_linear_second_layer(act_func(sharded_linear_first_layer(inp)))
+        sharded_output = aggregate_partial_tensor_list(sharded_output, self.rank)
+
+        # Run local computation
+        local_output = local_linear_second_layer(act_func(local_linear_first_layer(inp)))
+
+        # Verify
+        self.assertEqual(local_output, sharded_output)
+
+        # Compute loss and run backward pass.
+        local_output.sum().backward()
+        sharded_output.sum().backward()
+        local_grad_first = local_linear_first_layer.weight.grad
+        local_grad_second = local_linear_second_layer.weight.grad
+
+        # Verify that weights in both layers and biases in the sharded linear has non-None grad.
+        sharded_weight_first = sharded_linear_first_layer.weight.local_shards()[0].tensor
+        sharded_weight_second = sharded_linear_second_layer.weight.local_shards()[0].tensor
+        self.assertNotEqual(sharded_weight_first.grad, None)
+        self.assertNotEqual(sharded_weight_second.grad, None)
+        self.assertNotEqual(sharded_linear_first_layer.bias.grad, None)
+        self.assertNotEqual(sharded_linear_second_layer.bias.grad, None)
+
+        # Shard the local linear's weight grad so that we can compare.
+        dist.all_reduce(local_grad_first)
+        dist.all_reduce(local_grad_second)
+        (start_pos, chunk_size) = generate_local_weight_sharding_params_for_test(
+            local_linear_first_layer.weight, 0, TEST_GPU_NUM, spec[0], self.rank
+        )
+        local_grad_narrowed_first = local_grad_first.narrow(0, start_pos, chunk_size)
+        (start_pos, chunk_size) = generate_local_weight_sharding_params_for_test(
+            local_linear_second_layer.weight, 1, TEST_GPU_NUM, spec[1], self.rank
+        )
+        local_grad_narrowed_second = local_grad_second.narrow(1, start_pos, chunk_size)
+
+        # Test backward gradient calculation.
+        self.assertEqual(sharded_weight_first.grad, local_grad_narrowed_first)
+        self.assertEqual(sharded_weight_second.grad, local_grad_narrowed_second)
```

Comment:
Add a TODO here for optimizer step.

Reply:
Will rebase on top of PR: https://github.com/pytorch/pytorch/pull/69569